src/bots/liquidator.ts

```diff
@@ -82,17 +117,197 @@ export class LiquidatorBot implements Bot {
+			`    liquidatorPosition1: ${convertToNumber(
+				newPosition.baseAssetAmount,
+				BASE_PRECISION
+			).toString()}`
+		);
+
+		// calculate margin required to take over position
+		const worstCaseBaseAssetAmount =
+			calculateWorstCaseBaseAssetAmount(newPosition);
+		const worstCaseAssetValue = worstCaseBaseAssetAmount
+			.abs()
+			.mul(
+				this.clearingHouse.getOracleDataForMarket(
+					liquidateePosition.marketIndex
+				).price
+			)
+			.div(AMM_TO_QUOTE_PRECISION_RATIO.mul(MARK_PRICE_PRECISION));
+
+		const marketMarginRatio = new BN(
+			calculateMarketMarginRatio(market, worstCaseBaseAssetAmount, 'Initial')
+		);
+		const marginRequired = worstCaseAssetValue
+			.mul(marketMarginRatio)
+			.div(MARGIN_PRECISION);
+
+		const marginAvailable = liquidatorUser.getFreeCollateral();
+		logger.info(
+			`    marginRequired: ${convertToNumber(
+				marginRequired,
+				QUOTE_PRECISION
+			).toString()}`
+		);
+		logger.info(
+			`    marginAvailable: ${convertToNumber(
+				marginAvailable,
+				QUOTE_PRECISION
+			).toString()}`
+		);
+		logger.info(
+			`      enough collateral to liquidate??: ${marginAvailable.gte(
+				marginRequired
+			)}`
+		);
+
+		return marginAvailable.gte(marginRequired);
+	}
+
+	private calculateBaseAmountToLiquidate(
+		liquidatorUser: ClearingHouseUser,
+		liquidateePosition: UserPosition
+	): BN {
+		const oraclePrice = this.clearingHouse.getOracleDataForMarket(
+			liquidateePosition.marketIndex
+		).price;
+		const collateralToSpend = liquidatorUser
+			.getFreeCollateral()
+			.mul(MARK_PRICE_PRECISION)
+			.mul(this.MAX_POSITION_TAKEOVER_PCT_OF_COLLATERAL)
+			.mul(BASE_PRECISION);
+		const baseAmountToSpend = collateralToSpend.div(
+			oraclePrice
+				.mul(QUOTE_PRECISION)
+				.mul(this.MAX_POSITION_TAKEOVER_PCT_OF_COLLATERAL_DENOM)
+		);
+
+		if (baseAmountToSpend.gt(liquidateePosition.baseAssetAmount.abs())) {
+			return liquidateePosition.baseAssetAmount.abs();
+		} else {
+			return baseAmountToSpend;
+		}
+	}
+
+	/**
+	 * iterates over users in userMap and chekcs if they can be liquidated. If so, their positions are checked to find the
+	 * endangered position to liquidate.
+	 */
 	private async tryLiquidate() {
 		try {
 			for (const user of this.userMap.values()) {
 				const auth = user.getUserAccount().authority.toBase58();
 				const userKey = user.userAccountPublicKey.toBase58();
-				// console.log(`[${auth}: ${userKey}]`);
-				// console.log(`  leverage: ${convertToNumber(user.getLeverage(), TEN_THOUSAND).toString()}`);
 				const [canBeLiquidated, _marginRatio] = user.canBeLiquidated();
-				// console.log(`  canBeLiquidated: ${canBeLiquidated}, marginRatio: ${convertToNumber(marginRatio, TEN_THOUSAND).toString()}`);
-				// const bankLiabilityValue = user.getBankLiabilityValue();
-				// console.log(`  bankLiabilityValue: ${convertToNumber(bankLiabilityValue, QUOTE_PRECISION).toString()}`);
 
 				if (canBeLiquidated) {
 					logger.info(`liquidating ${auth}: ${userKey}...`);
```

Comment:
not sure if you need to call since the accounts will be polled in the background?
```
					this.clearingHouse.fetchAccounts();
					this.clearingHouse.getUser().fetchAccounts();
```